libr/util/charset.c

```diff
@@ -113,24 +113,36 @@ R_API size_t r_charset_encode_str(RCharset *rc, ut8 *out, size_t out_len, const
 
 // assumes out is as big as in_len
 R_API size_t r_charset_decode_str(RCharset *rc, ut8 *out, size_t out_len, const ut8 *in, size_t in_len) {
-	char k[32];
 	char *o = (char*)out;
-	int i;
-	for (i = 0; i < in_len; i++) {
-		ut8 ch_in = in[i];
-
-		//zero terminate the string
-		snprintf (k, sizeof (k), "%c", ch_in);//snprintf (k, sizeof (k), "0x%02x", ch_in);
-		char *v = sdb_get (rc->db_char_to_hex, k, 0);
-		memmove(v, v+2, strlen (v));
-
-		//convert to ascii
-		char str_hx[32];
-		snprintf (str_hx, sizeof (str_hx), "%c", (char) strtol( v, 0, 16));
-		const char *ret = r_str_get_fail (str_hx, "?");
-		strcpy (o, ret);
-
-		o += strlen (o);
+	int j;
+	ut8 str[320];
+	size_t last_char_size;
+
+	size_t cur;
+	for (cur = 0; cur < in_len; ) {
+		memcpy (str, in, sizeof (str));
+		for (j =( in_len+1 ); j > 0; j--) {
+			str[j] = '\0';
+			//zero terminate the string
+			char *v = sdb_get (rc->db_char_to_hex, (char *) str+cur, 0);
+
+			if (v) {
+				memmove (v, v+2, strlen (v));
+
+				//convert to ascii
+				char str_hx[320];
+				snprintf (str_hx, sizeof (str_hx), "%c", (char) strtol(v, 0, 16));//in the future handle multiple chars output
+				const char *ret = r_str_get_fail (str_hx, "?");
+
+				//concatenate
+				strcpy (o, ret);
+				o += strlen (o);
```

Comment:
o = strcpy (o, ret);

Reply:
its always 1 for now and strlen shouldnt be needed when implemented multichar support

Reply:
`o = strcpy (o, ret);` writes only the first char. For example in `w <RIVAL>abc`, only <RIVAL> is printed. It is a great bug.